crates/signer/src/signer.rs

```diff
@@ -0,0 +1,268 @@
+use crate::web3_signer::Web3Signer;
+use alloy_primitives::Address;
+use alloy_signer_aws::{AwsSigner, AwsSignerError};
+use alloy_signer_local::PrivateKeySigner;
+use eth_keystore::decrypt_key;
+use std::path::Path;
+use thiserror::Error;
+use url::Url;
+
+/// Represents the input params to create a signer
+#[derive(Debug)]
+pub enum Config {
+    /// Hexadecimal private key
+    PrivateKey(String),
+    /// Keystore path and password
+    Keystore(String, String),
+}
+
+/// Possible errors raised in signer creation
+#[derive(Error, Debug)]
+pub enum SignerError {
+    #[error("invalid private key")]
+    InvalidPrivateKey,
+    #[error("invalid keystore password")]
+    InvalidPassword,
+    #[error("invalid address")]
+    InvalidAddress,
+    #[error("invalid url")]
+    InvalidUrl,
+}
+
+impl Config {
+    /// Creates a signer from given config.
+    pub fn signer_from_config(c: Config) -> Result<PrivateKeySigner, SignerError> {
+        // TODO: check chain id to select signer
+        match c {
+            Config::PrivateKey(key) => key
+                .parse::<PrivateKeySigner>()
+                .map_err(|_| SignerError::InvalidPrivateKey),
+            Config::Keystore(path, password) => {
+                let keypath = Path::new(&path);
+                let private_key =
+                    decrypt_key(keypath, password).map_err(|_| SignerError::InvalidPassword)?;
+                PrivateKeySigner::from_slice(&private_key)
+                    .map_err(|_| SignerError::InvalidPrivateKey)
+            }
+        }
+    }
+    /// Creates a signer from a key ID in AWS Key Management Service
+    pub async fn aws_signer(
+        key_id: String,
+        chain_id: Option<u64>,
+        client: aws_sdk_kms::Client,
+    ) -> Result<AwsSigner, AwsSignerError> {
+        AwsSigner::new(client, key_id, chain_id).await
+    }
+
+    /// Creates a signer that uses the Web3Signer JSON-RPC API
+    pub fn web3_signer(endpoint: String, address: Address) -> Result<Web3Signer, SignerError> {
+        let url: Url = endpoint.parse().map_err(|_| SignerError::InvalidUrl)?;
+        Ok(Web3Signer::new(address, url))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::Config;
+    use alloy_consensus::{SignableTransaction, TxLegacy};
+    use alloy_network::{TxSigner, TxSignerSync};
+    use alloy_node_bindings::Anvil;
+    use alloy_primitives::hex_literal::hex;
+    use alloy_primitives::{address, bytes, keccak256, Address, U256};
+    use alloy_signer::Signature;
+    use alloy_signer_local::PrivateKeySigner;
+    use aws_config::{BehaviorVersion, Region, SdkConfig};
+    use aws_sdk_kms::types::KeyMetadata;
+    use aws_sdk_kms::{
```

Comment:
join this 2 includes inside the second one

Reply:
Done